### ISSUE TITLE
No support for TINYINT(1)?

### ISSUE THREAD

**Issue by gtrias** [dolthub/dolt#6682]
I'm experiencing type errors in my Rails application when trying to connect it to `dolt sql-server`. The MySQL adapter in Rails can connect to the database, but the issue arises when I load the application. I'm receiving type errors due to ActiveRecord returning an integer instead of a boolean for certain columns that are supposed to be boolean.

Here's an example error message:

```
[d099cd1a-e428-4eb7-96c9-4bf67c87304c] TypeError (Parameter 'enabled': Can't set Features::Entities::Feature.enabled to 1 (instance of Integer) - need a T::Boolean
Caller: /cache/bundle/gems/sorbet-runtime-0.5.10974/lib/types/private/methods/call_validation.rb:256
```

After researching, I believe the issue is related to `dolt` not considering the width of `tinyint`.

For example, in my SQL dump, the `features` table is defined as follows (note the `tinyint(1)`):

```sql
CREATE TABLE `features` (
  `id` int NOT NULL AUTO_INCREMENT,
  `company_id` int NOT NULL,
  `name` varchar(255) NOT NULL,
  `enabled` tinyint(1) NOT NULL,
  `created_at` datetime(6) DEFAULT NULL,
  `updated_at` datetime(6) DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `index_features_on_company_id` (`company_id`),
  KEY `index_features_on_name_and_enabled` (`name`,`enabled`),
  CONSTRAINT `fk_rails_e6b730face` FOREIGN KEY (`company_id`) REFERENCES `companies` (`id`) ON DELETE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=40 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
```

To import this dump, I started a `dolt sql-server` instance and connected to it using a regular `mysql` client. Then, I ran `source my-dump.sql` to import the database.

Upon analyzing the table schema using `doltlab`, I noticed that `tinyint` doesn't have any size specified in `dolt`. Here's the schema reported by `doltlab`:

```sql
CREATE TABLE `features` (
  `id` int NOT NULL AUTO_INCREMENT,
  `company_id` int NOT NULL,
  `name` varchar(255) NOT NULL,
  `enabled` tinyint NOT NULL,
  `created_at` datetime(6) DEFAULT NULL,
  `updated_at` datetime(6) DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `index_features_on_company_id` (`company_id`),
  KEY `index_features_on_name_and_enabled` (`name`,`enabled`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

Is there a solution to make `dolt` consider `TINYINT(1)` properly?

**Comment by macneale4** [comment]
I can verify that we aren't tracking the tinyint width:

```
db2> create table foo (enabled tinyint(1));
db2> show create table foo;
+-------+------------------------------------------------------------------+
| Table | Create Table                                                     |
+-------+------------------------------------------------------------------+
| foo   | CREATE TABLE `foo` (                                             |
|       |   `enabled` tinyint                                              |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_bin |
+-------+------------------------------------------------------------------+
1 row in set (0.00 sec)

db2> create table foo2 (enabled tinyint(2));
db2> show create table foo2;
+-------+------------------------------------------------------------------+
| Table | Create Table                                                     |
+-------+------------------------------------------------------------------+
| foo2  | CREATE TABLE `foo2` (                                            |
|       |   `enabled` tinyint                                              |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_bin |
+-------+------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by zachmu** [comment]
Dolt doesn't store the length parameter for integer types because they don't affect the storage. They're display width hints to the client application.

https://dev.mysql.com/doc/refman/8.0/en/numeric-type-attributes.html

You aren't the first person to notice, but we haven't had any reports of it causing a problem -- this is the first instance we've heard of this causing a compatibility issue. That's an argument in favor of storing these length params.

**Comment by fulghum** [comment]
Hey @gtrias, thanks for opening this issue to let us know about the problem with Rails and `tinyint` 🙏 

I've made a simple repro with Rails and am just starting to dig into the issue to see how we can fix it. In the meantime, I wanted to share a workaround I tested out, just in case that's helpful while we're working on the fix. 

I'm by no means a Ruby or Rails expert, so please feel free to correct me if you know a better workaround! Here's how I was able to change my ActiveRecord configuration to treat a `tinyint` field as a boolean:
```ruby
class Test1 < ApplicationRecord
  self.attribute :b1, ActiveRecord::Type::Boolean.new
end
```

We'll follow up soon with more details on the fix as we dig deeper!

**Comment by fulghum** [comment]
Quick update... ActiveRecord is using the `SHOW FULL FIELDS FROM <tablename>;` statement to load this information. I hacked a quick test together to hardcode returning `tinyint(1)` and confirmed that once ActiveRecord sees that, it treats the field as a boolean, without requiring the workaround above. 

Now that we've confirmed what needs to change, I'll get started on storing and returning the display width for `tinyint` types.

**Comment by fulghum** [comment]
I've got PRs out for this fix and have confirmed locally that this change allows ActiveRecord to correctly map `TINYINT(1)` columns in Dolt to boolean fields in Ruby. I've also opened issue https://github.com/dolthub/dolt/issues/6689 as a follow up for us to add more ORM integration tests for ActiveRecord to get better test coverage there. 

I'm currently waiting for the Dolt CI tests to pass, then I'll get these changes reviewed. I expect we'll be able to get this fix released tomorrow for you.

**Comment by gtrias** [comment]
Thank you very much for this quick response!!! Looking forward to running our rails application with dolt :)

**Comment by fulghum** [comment]
> The problem with this approach is that I should change 390 booleans that we have in our database 😅
> I think I will wait for dolt to support tinyint width

Oof! That would be a lot to update! 😳 I've got the PRs cleaned up, the tests are passing, and I just pushed them over to review from a teammate. We're on track to get the fix released today for you.

**Comment by fulghum** [comment]
Code changes are checked in for this fix. After they finish the last set of CI checks, I'll cut a release and let you know when the fix is available to pick up.

**Comment by fulghum** [comment]
@gtrias – This fix is now available in [Dolt 1.16.2](https://github.com/dolthub/dolt/releases/tag/v1.16.2). 

Thank you for taking the time to file this issue so we could get this fixed! 🙏  Let us know if you run into any other problems with Dolt and we'll be happy to jump on them.

**Comment by gtrias** [comment]
Wow! I'm going to test it right now!! I'm still amazed on how fast you arranged all this :clap:
